components/compositing/compositor.rs

```diff
@@ -862,20 +867,69 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
         }
     }
 
+    /// Create a WebRender display list for the root pipeline in the given transaction.
+    /// This display list contains a full-screen iframe with the root content pipeline
+    /// with any pinch zoom transform applied to it.
+    fn set_root_display_list(&self, transaction: &mut Transaction) {
+        let root_content_pipeline = match self.root_content_pipeline.id {
+            Some(id) => id,
+            None => return,
+        };
+
+        let root_pipeline = WebRenderPipelineId(0, 0);
+        transaction.set_root_pipeline(root_pipeline);
+
+        let mut builder = webrender_api::DisplayListBuilder::new(root_pipeline);
+        let viewport_size = LayoutSize::new(
+            self.embedder_coordinates.get_viewport().width() as f32,
+            self.embedder_coordinates.get_viewport().height() as f32,
+        );
+        let zoom_factor = self.pinch_zoom_level();
+        let viewport_rect = LayoutRect::new(LayoutPoint::zero(), viewport_size);
+        let zoom_reference_frame = builder.push_reference_frame(
+            LayoutPoint::zero(),
```

Comment:
This not strictly related to this PR, but something I noticed when working on the gesture support for Android -  I understand the current webrender's pinch-zoom api doesn't accept the zoom origin. Would this new implementation allow pinch-zooming from a given point or will it still be from top-left corner?

Reply:
This should be fixable in a followup change, but requires answering some questions about how to differentiate between scrolling and zoom in on content.